pypowerwall/scan.py

```diff
@@ -35,7 +48,50 @@ def getmyIP():
     s.close()
     return r
 
-def scan(color=True, timeout=0.4, ip=None):
+def scanIP(color, timeout, addr):
+    global discovered, firmware
+    global bold, subbold, normal, dim, alert, alertdim
+
+    if(color == False):
+        # Disable Terminal Color Formatting
+        bold = subbold = normal = dim = alert = alertdim = ""
+
+    host = dim + '\r      Host: ' + subbold + '%s ...' % addr + normal
+    print(host, end='')
+    a_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    a_socket.settimeout(timeout)
+    location = (str(addr), 443)
+    result_of_check = a_socket.connect_ex(location)
+    if result_of_check == 0:
```

Comment:
I found a pre-existing bug! Your change is lighting fast, but I noticed that every once in a while it wouldn't find my Powerwall. I have a LOT of devices on my network including several with port 443 open and the number it found would vary. 

I looked it up and connect_ex() can return errno.EAGAIN which indicates that some blocking event occurred so instead of waiting, it just hands back errno.EAGAIN.  I add a print(result_of_check) and saw several EAGAIN errors. I made a test change to essential honor errno.EAGAIN by "trying again" after minor delay:

```python
    while True:
        result_of_check = a_socket.connect_ex(location)
        if not result_of_check == socket.errno.EAGAIN:
            break
        time.sleep(0.1)
    if result_of_check == 0:
```

With that change, it gives consistent results. I tested on MacOS and Linux.  I need to check WinOS.

Reply:
Great find!